// prisma-dbml-generator/dist/generator/project.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getProjectOptions = exports.generateProject = void 0;
const fs_1 = require("fs");
const { readFile } = fs_1.promises;
function generateProject({ name, databaseType, note, isMd = false, }) {
    const projectNote = isMd
        ? `'''\n` +
            `    ${note
                .replace(/\n/g, '\n    ')
                .replace(/(\n\s+\n)/g, '\n\n')
                .replace(/\s+$/g, '')}\n  '''`
        : `'${note}'`;
    const project = [
        `Project ${name} {\n` +
            `  database_type: '${databaseType}'\n` +
            `  Note: ${projectNote}\n}`,
    ];
    return name ? project : [];
}
exports.generateProject = generateProject;
async function getProjectOptions({ projectName, projectDatabaseType, projectNote, projectNotePath, }) {
    let projectNoteMd = '';
    if (projectNotePath) {
        const fullPath = `${process.cwd()}/${projectNotePath}`;
        try {
            projectNoteMd = await readFile(fullPath, 'utf-8');
        }
        catch (e) {
            console.log(`❌ Error: project note markdown file not found: ${fullPath}`);
        }
    }
    return {
        name: projectName && `"${projectName}"`,
        databaseType: projectDatabaseType || '',
        note: projectNoteMd || projectNote || '',
        isMd: projectNoteMd !== '',
    };
}
exports.getProjectOptions = getProjectOptions;
