// prisma-dbml-generator/dist/keywords.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrismaScalars = exports.DBMLKeywords = void 0;
var DBMLKeywords;
(function (DBMLKeywords) {
    DBMLKeywords["Default"] = "default";
    DBMLKeywords["Increment"] = "increment";
    DBMLKeywords["Note"] = "note";
    DBMLKeywords["NotNull"] = "not null";
    DBMLKeywords["Table"] = "Table";
    DBMLKeywords["Unique"] = "unique";
    DBMLKeywords["Pk"] = "pk";
    DBMLKeywords["Indexes"] = "indexes";
})(DBMLKeywords = exports.DBMLKeywords || (exports.DBMLKeywords = {}));
var PrismaScalars;
(function (PrismaScalars) {
    PrismaScalars["String"] = "String";
    PrismaScalars["Boolean"] = "Boolean";
    PrismaScalars["Int"] = "Int";
    PrismaScalars["Float"] = "Float";
    PrismaScalars["DateTime"] = "DateTime";
    PrismaScalars["Json"] = "Json";
    PrismaScalars["BigInt"] = "BigInt";
    PrismaScalars["Decimal"] = "Decimal";
    PrismaScalars["Bytes"] = "Bytes";
})(PrismaScalars = exports.PrismaScalars || (exports.PrismaScalars = {}));
